File: leetcode/src/com/leetcode/all/tree/No103_zigzagLevelOrder.java
package com.leetcode.all.tree;

import java.util.*;

/**
 * 给定一个二叉树，返回其节点值的锯齿形层序遍历。（即先从左往右，再从右往左进行下一层遍历，以此类推，层与层之间交替进行）。
 * 给定二叉树 [3,9,20,null,null,15,7]
 * 3
 * / \
 * 9  20
 * /  \
 * 15   7
 * 输出：
 * [
 * [3],
 * [20,9],
 * [15,7]
 * ]
 *
 * @author Bi11ions
 * @description TODO
 * @date 2021/3/11 17:03
 */
public class No103_zigzagLevelOrder {

    public List<List<Integer>> zigzagLevelOrder(TreeNode root) {
        if (null == root) {
            return new ArrayList<>(0);
        }

        List<List<Integer>> resultList = new ArrayList<>();
        Queue<TreeNode> queue = new LinkedList<>();
        queue.offer(root);
        boolean reverse = true;
        while (!queue.isEmpty()) {
            int size = queue.size();
            Deque<Integer> subList = new LinkedList<>();
            for (int i = 0; i < size; i++) {
                TreeNode node = queue.poll();
                if (reverse) {
                    subList.offerLast(node.val);
                } else {
                    subList.offerFirst(node.val);
                }

                if (null != node.left) {
                    queue.offer(node.left);
                }

                if (null != node.right) {
                    queue.offer(node.right);
                }
            }

            resultList.add(new ArrayList<>(subList));
            reverse = !reverse;
        }

        return resultList;
    }
}

File: leetcode-go/src/main/no103_zigzagLevelOrder.go
package main

// 二叉树的锯齿形层次遍历
func zigzagLevelOrder(root *TreeNode) [][]int {
	if nil == root {
		return [][]int{}
	}

	var resultList [][]int
	queue := []*TreeNode{root}
	for level := 0; len(queue) > 0; level++ {
		var values []int
		q := queue
		queue = nil
		for _, node := range q {
			values = append(values, node.Val)
			if node.Left != nil {
				queue = append(queue, node.Left)
			}

			if node.Right != nil {
				queue = append(queue, node.Right)
			}
		}

		// 本质上和层序遍历一样，我们只需要把奇数层的元素翻转即可
		if level%2 == 1 {
			for i, n := 0, len(values); i < n/2; i++ {
				values[i], values[n-1-i] = values[n-1-i], values[i]
			}
		}

		resultList = append(resultList, values)
	}

	return resultList
}
